Query Loop: Default to querying posts when on singular content

mikachan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
An attempt to try and fix the issue described in https://github.com/WordPress/gutenberg/issues/64746.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
To align the default/inherit query results between the Editor and the front end, and also hopefully make this block more intuitive and easier to understand.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This adds a check for `is_singular()` to check if the block has been added to a single post or page (i.e. within content), and if so, defaults to querying the `post` post type.

It also hides the Query Type control when editing a single page/post/CPT, as these blocks will always default to a custom query type that queries posts. The Query Type control is still used when editing a template, where it makes more sense to default to inheriting the global query.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Insert a Query Loop block inside a single page, post, or CPT.
2. Ensure that the Default query ("inherit") shows posts from the 'post' post type in both the Editor and the front-end.
3. Insert a Query Loop block inside a template.
4. Ensure that the "Query Type" control is only shown when editing a template and that it is hidden when editing a single page, post, or CPT.

## Screenshots or screencast <!-- if applicable -->

### Front-end (posts should now render by default)

| Before | After |
| ----- | ----- |
| <img width="690" alt="image" src="https://github.com/user-attachments/assets/99e88f9b-7334-41eb-ada3-a4562e1c6356"> | <img width="690" alt="image" src="https://github.com/user-attachments/assets/921656c9-3d95-412a-93b7-0ac4e8887219"> |

### Editor (shouldn't be any change)

| Before | After |
| ----- | ----- |
| <img width="1433" alt="image" src="https://github.com/user-attachments/assets/df875a12-9c37-48ff-8cdd-55a8f5d10566"> | <img width="1433" alt="image" src="https://github.com/user-attachments/assets/f9aefa1b-0ab4-4b6f-9f32-aba34b1efdcb"> |

### Query Type control changes (currently, the Query Type control is always shown)

| Editing a template | Editing content (e.g. page) |
| -------- | ------- |
| <img width="675" alt="image" src="https://github.com/user-attachments/assets/0749b362-3f1b-476f-b749-c59fcb605b47"> | <img width="675" alt="image" src="https://github.com/user-attachments/assets/1ae1666c-b0e4-4412-b73f-964f3ba999fb"> |

mikachan committed:
Fix linter warnings

mikachan committed:
Replace the postType if current post type is different

fabiankaegy commented on packages/block-library/src/query/edit/query-content.js:
```diff
@@ -46,10 +46,14 @@ export default function QueryContent( {
+		// @wordpress/block-library should not depend on @wordpress/editor.
+		// However, we need to get the current post type in order to inherit the correct query.
+		// eslint-disable-next-line @wordpress/data-no-store-string-literals
+		const { getCurrentPostType } = select( 'core/editor' );
```
If we need to access the current postType we should be able to get it via the block context by setting `usesContext` to include `postType`.

However I'm not sure this approach is solving the problem described 🤔 

Instead of showing the posts of the current post type I do think the default state for the inherit option should still be that is always renders posts of the posts posts type. 

The issue is that today it doesn't do that when the query loop is inserted inside content rather than inside a template. 

Rather than this I think it might be a better route to add a check in PHP to check whether something like whether the inherit option is set to true and the `is_singular` check resolves to true. 

In that case we change the default params to set the post type to `post`

mikachan replied:
Thanks @fabiankaegy! 

> If we need to access the current postType we should be able to get it via the block context by setting usesContext to include postType.

Ah thank you, that's helpful.

> Instead of showing the posts of the current post type I do think the default state for the inherit option should still be that is always renders posts of the posts posts type.

I agree - I was focused on making the Editor and the front end consistent, and assumed the front end was correct. But now I think it makes more sense for the query loop to default to querying posts, as this seems to be the safest assumption for most users. I've tried out `is_singular()` here: https://github.com/WordPress/gutenberg/pull/65067/commits/51f93838592dc2456332967e847985985b1f7bd1

This new behavior also matches the block description:

![image](https://github.com/user-attachments/assets/16dc16b2-8e2d-4754-8e74-645c9a6e24d8)

mikachan committed:
Remove currentPostType logic

mikachan committed:
Default to posts if is_singular

mikachan committed:
Update test_rendering_post_template_with_main_query_loop_already_started test

mikachan committed:
Add test for query loop not inside a singular query

mikachan committed:
Revert changes to QueryContent

creativecoder commented:
I'm not sure that it always makes sense to default to the `post` post type.

If you're on a single page, I can see the logic of doing that. However, if I'm on a single template for a custom post type, defaulting to posts is unexpected--most of the time, I'd expect to see a list of posts _of the same_ post type.

What do you think of defaulting the query block to the post type that matches the currently viewed post?

jameskoster commented:
We should check how the `Default` setting behaves in different contexts. For instance what happens if that query appears in search results, or a category archive? Or if there's a custom query in a template to load in the content of a post/page that contains such a `default` Query.

"Default" is potentially a bit misleading because it suggests the _global default_, which isn't how the option works. We can _force_ it to work that way, but won't that involve adding a lot of logic to account for scenarios like the ones I mentioned above?

As an alternative, I still think it might be worth exploring the idea of forcing Query Loops inside Content blocks to be custom, and hiding the 'Query Type' control altogether. The values for controls in the custom query set up could inherit from the global query as a default.

fabiankaegy commented:
Yeah this should never override a global query if one exists. If you are on an archive of a custom post type, or on the search template it should just inherit the global as a default. 

The crux of the issue is that created is that when there is no global query, like on single templates. It should default to posts.

creativecoder commented:
> The crux of the issue is that created is that when there is no global query, like on single templates. It should default to posts.

@fabiankaegy That makes sense to me for `page` and `post` post types.

On single templates for custom post types (like a custom book, movie, news, or location post type), would you still expect the query to default to the `post` post type? For me that would be unexpected--I would expect it to show me other posts from the same custom post type that I'm currently viewing.

fabiankaegy commented:
@creativecoder from my perspective I would prefer the predictability of it defaulting to posts on all single templates unless you choose custom.  

We simply don’t have enough insights into what a CPT is used for to determine whether it should query other items. 

And even today when you insert a query loop on a single page of some CPT it displays the most recent posts in the editor. It just doesn’t on the frontend today.

mikachan committed:
Merge branch 'trunk' into fix/query-loop-posttype

mikachan commented:
Thanks for the input here! I got myself a little tangled trying to address how the query loop handles CPT in this PR, especially in the Editor. It would be ideal for the query loop to work out what the best CPT is based on the context of where it is inserted, but at the moment I don't believe we have the underlying logic required to do this in the Editor. 

My vote would be to work on this logic in follow-up PRs (if needed), and for this PR to focus on fixing the inconsistency between the Editor and the front-end by always defaulting to displaying posts. At least this improves the current UX for the block.

If the block always defaults to 'post' posts, I think this would also be aligned with @jameskoster's idea here:

> As an alternative, I still think it might be worth exploring the idea of forcing Query Loops inside Content blocks to be custom, and hiding the 'Query Type' control altogether. The values for controls in the custom query set up could inherit from the global query as a default.

If the controls for the query are always visible in the block settings like the current 'Custom' view, it would be clear to users that 'post' is the default post type:

<img width="269" alt="image" src="https://github.com/user-attachments/assets/8fc6beca-120a-4812-a8dc-d2233c6cc612">

mikachan committed:
Show query type control only when on a template

creativecoder commented:
> from my perspective I would prefer the predictability of it defaulting to posts on all single templates unless you choose custom.

Even though I find that a bit unexpected for custom post types, I think it's a reasonable default.

> My vote would be to work on this logic in follow-up PRs (if needed), and for this PR to focus on fixing the inconsistency between the Editor and the front-end by always defaulting to displaying posts. 

Agreed that sounds like a good plan for this PR.

mikachan commented:
I think this PR is already fixing the inconsistency that is currently happening with this block by updating the query to default to the 'post' post type when `inherit` is true and the block is used on a single post, page or CPT (i.e. content). This will be overwritten if the user chooses a different query using the block settings.

With the latest commit, I've also tried out @jameskoster's idea of hiding the Query Type control when not on a template: https://github.com/WordPress/gutenberg/pull/65067/commits/58c0956883300de8716acd6f252ec1ee82a73fc0. Let me know what you think, or I'm happy to split this into a separate PR if folks prefer.

| Editing a template | Editing content (e.g. page) |
| -------- | ------- |
| <img width="675" alt="image" src="https://github.com/user-attachments/assets/0749b362-3f1b-476f-b749-c59fcb605b47"> | <img width="675" alt="image" src="https://github.com/user-attachments/assets/1ae1666c-b0e4-4412-b73f-964f3ba999fb"> |

mikachan committed:
Merge branch 'trunk' into fix/query-loop-posttype

# Conflicts:
#	packages/block-library/src/query/edit/inspector-controls/index.js

mikachan committed:
Move template logic to QueryInspectorControls

mikachan committed:
Ensure inherit value is updated when not in a template

mikachan committed:
Update comment

mikachan committed:
Rename showDefaultControl to isTemplate

fabiankaegy reviewed (approved):
This teste well for me and I'm in favor of getting this in for 6.7. yes it is a bug so we have more time. But I think we can even merge it now and get it in before the deadline.

jameskoster commented:
What am I doing wrong, I still see the option when editing a page? 🤔 

<img width="1227" alt="Screenshot 2024-09-18 at 17 02 12" src="https://github.com/user-attachments/assets/779b2b9d-23aa-460e-9f7a-4d50d01ea28f">

mikachan committed:
Get postType from context

mikachan committed:
Add a check for singular content based on available post type

mikachan commented:
> What am I doing wrong, I still see the option when editing a page?

Good spot - I think it's because you're testing a page within the Editor, rather than via wp-admin -> Pages. I've added a check for editing singular content within the Editor which I think should cover this. Thanks @jameskoster!

mikachan committed:
Move inherit reset to a useEffect

jameskoster commented:
That fixed it :) 

Y'all may consider this an edge case not worth catering to, but how do you think we should handle Query Loops that have set `inherit: true`, either via the code editor, or when pasted into the editor? I noticed some quirks in the current UX, best illustrated with a video...

https://github.com/user-attachments/assets/d6256597-10eb-44c9-9abe-f25c98ef0ae2

* Manually changing the `inherit` value marks the page as dirty; I can save that change.
* When I select Query Loop `inherit` seems to be reset to `false` behind the scenes.

---

I wonder if we should consider disabling the Query Type option in some templates like 404 as well. It doesn't make much sense there given there's no query to inherit.

mikachan committed:
Move isTemplate logic to QueryContent

mikachan committed:
Fix lint warnings

mikachan commented on packages/block-library/src/query/edit/query-content.js:
```diff
@@ -81,6 +93,10 @@ export default function QueryContent( {
 	// Changes in query property (which is an object) need to be in the same callback,
 	// because updates are batched after the render and changes in different query properties
 	// would cause to override previous wanted changes.
+	const updateQuery = useCallback(
+		( newQuery ) => setAttributes( { query: { ...query, ...newQuery } } ),
+		[ query, setAttributes ]
+	);
```
This was from fixing a lint warning.

mikachan commented:
> Manually changing the inherit value marks the page as dirty; I can save that change.
> When I select Query Loop inherit seems to be reset to false behind the scenes.

Ah, I think I was checking for `isTemplate` too late. This check automatically sets `inherit` to `false` when editing a page/post/CPT. I've moved this logic higher up into `QueryContent` in https://github.com/WordPress/gutenberg/pull/65067/commits/47c84a7faddbc923d7a890f6d792eb309825fe5f and this _seems_ to work better, as the value seems to be set to false more reliably when switching between editing modes. It also means we can use `markNextChangeAsNonPersistent` after resetting the value, as @fabiankaegy mentioned.

I also fixed some lint warnings in `QueryContent` in https://github.com/WordPress/gutenberg/pull/65067/commits/7447863b2d2d06b43dfdcfa29e16d6d9971cb93e, since I was editing the code around them.

> I wonder if we should consider disabling the Query Type option in some templates like 404 as well. It doesn't make much sense there given there's no query to inherit.

Yeah that sounds like a good plan, although I agree with @fabiankaegy in that this might be better in a follow-up.

This is ready for another review. Thanks both 🙏

noisysocks commented:
About to publish GB 19.3 and I see an approval here so just going to merge 😇

dlh01 commented:
Please pardon my quick, after-the-fact comment, but I wanted to flag the use of `query_posts()` in this PR, as I'm not sure that it's going to have the intended effect.

The comment just above that block of code begins:

> If already in the main query loop, duplicate the query instance to not tamper with the main instance.

followed by a `clone` of the query object. 

But `query_posts()` does "tamper" with the main query instance [by replacing it](https://github.com/WordPress/wordpress-develop/blob/bf09fe506620678fb82c3b872309edda0ed8ce61/src/wp-includes/query.php#L79-L100). For example, I think if that call to `query_posts()` runs, any subsequent calls to `is_singular()` will return `false` because the global `$wp_query` has been replaced with the query for posts.

mikachan commented:
Ah I see, thanks @dlh01! I've just tried replacing `query_posts( array( 'post_type' => 'post' ) )` with `$query->set( 'post_type', 'post' )`, which seems to work well in my testing. I've opened a PR here: https://github.com/WordPress/gutenberg/pull/65483. Does this sound like a good alternative?

dlh01 commented:
Thanks for taking a look, @mikachan!

The change in #65483 looks OK to me as far as protecting the main query instance, but I'm not certain that it's actually having an effect. By the time the post template block is rendering, that query object has already executed, and changing the query parameters at that point (via `set()`) won't affect the query's results.

Taking a step back: If I'm following correctly, the changes in this PR have made it so that if a query block is inserted into post content, that query will always use `inherit: false`. If so, and if the query blocks being used for testing are affected by that logic, wouldn't they bypass the block of code that was altered here and in #65483? I would have thought that the `$use_global_query` variable would be `false`. Perhaps there wasn't any change needed on the PHP side at all now — the `query_posts()` call could be removed, and that's it?

mikachan commented:
> By the time the post template block is rendering, that query object has already executed, and changing the query parameters at that point (via set()) won't affect the query's results.

I think I tested using `set()` when I first started working on this, and I saw the same results, so I couldn't understand why it was working now!

> Perhaps there wasn't any change needed on the PHP side at all now — the query_posts() call could be removed, and that's it?

Yep, looks like this is the case. The forceful `inherit` change was all that was needed. I've removed the `is_singular()` check as part of the [follow-up PR](https://github.com/WordPress/gutenberg/pull/65483). Thanks so much for your help and explanations 🙇

ellatrix commented on packages/block-library/src/query/edit/query-content.js:
```diff
@@ -41,13 +42,24 @@ export default function QueryContent( {
+				select( coreStore ).__experimentalGetTemplateForLink()?.type;
+			const isInTemplate = 'wp_template' === currentTemplate;
+			const isInSingularContent = postType !== undefined;
+			return isInTemplate && ! isInSingularContent;
```
Why was `isInSingularContent` not enough? What conditions do I need to test?

ellatrix commented on packages/block-library/src/query/edit/query-content.js:
```diff
@@ -41,13 +42,24 @@ export default function QueryContent( {
+	const isTemplate = useSelect(
+		( select ) => {
+			const currentTemplate =
+				select( coreStore ).__experimentalGetTemplateForLink()?.type;
```
This is not a valid call, it's missing the `link` argument. Without this argument it's essentially checking wither we are in the site editor or not, because in the site editor this will always load the front page template, and in any other editor this will load nothing.

mikachan replied:
Thanks for explaining this. Just closing the loop here that this was fixed in https://github.com/WordPress/gutenberg/pull/65820.
